File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ReaderSettingsFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.PreferencesClient
import com.nexussoft.verbum.models.ReadingMode
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.runEffect
import com.nexussoft.verbum.common.arch.with

/** Reader preferences. Persists on change; the parent mirrors the value into the reader. */
object ReaderSettingsFeature {
    data class State(val textScale: ReaderTextScale, val readingMode: ReadingMode = ReadingMode.PAGES, val focusMode: Boolean = false)

    sealed interface Action {
        data class ModeChanged(val mode: ReadingMode): Action
        data class FocusChanged(val enabled: Boolean): Action
        data class TextScaleChanged(val scale: ReaderTextScale) : Action
    }

    fun reducer(preferences: PreferencesClient): Reducer<State, Action> = Reducer { state, action ->
        when (action) {
            is Action.ModeChanged -> state.copy(readingMode=action.mode).with(runEffect { preferences.setString(ChapterReaderFeature.MODE_KEY,action.mode.name) })
            is Action.FocusChanged -> state.copy(focusMode=action.enabled).with(runEffect { preferences.setString(ChapterReaderFeature.FOCUS_KEY,action.enabled.toString()) })
            is Action.TextScaleChanged -> state.copy(textScale = action.scale).with(
                runEffect { preferences.setString(ReaderTextScale.PREFERENCE_KEY, action.scale.name) },
            )
        }
    }
}

File: VerbumKit/Sources/Features/Scripture/ReaderSettingsFeature.swift
import ComposableArchitecture
import Models

/// Reader preferences. Writes go straight to shared state so the chapter
/// behind the sheet re-renders as the user picks.
@Reducer
public struct ReaderSettingsFeature {
    @ObservableState
    public struct State: Equatable {
        @Shared(.readerTextScale) public var textScale
        @Shared(.readingMode) public var readingMode
        @Shared(.readerFocusMode) public var focusMode
        public init() {}
    }

    public enum Action: Equatable {
        case textScaleChanged(ReaderTextScale)
        case readingModeChanged(ReadingMode)
        case focusModeChanged(Bool)
    }

    public init() {}

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .readingModeChanged(let mode): state.$readingMode.withLock { $0 = mode }; return .none
            case .focusModeChanged(let enabled): state.$focusMode.withLock { $0 = enabled }; return .none
            case .textScaleChanged(let scale):
                state.$textScale.withLock { $0 = scale }
                return .none
            }
        }
    }
}
